optimade/adapters/structures/__init__.py

```diff
@@ -0,0 +1,89 @@
+# pylint: disable=undefined-variable
+from typing import Union
+
+from optimade.models import StructureResource
+
+from .aiida import get_aiida_structure_data
+from .ase import get_ase_atoms
+from .cif import get_cif
+from .proteindatabank import get_pdb, get_pdbx_mmcif
+from .pymatgen import get_pymatgen_structure
+
+
+__all__ = ("Structure",)
+
+
+class Structure:
+    """Lazy structure converter
+    :param structure: JSON OPTIMADE single structures resource entry.
+    """
+
+    _structure_type_converters = {
+        "aiida_structuredata": get_aiida_structure_data,
+        "ase": get_ase_atoms,
+        "cif": get_cif,
+        "pdb": get_pdb,
+        "pdbx_mmcif": get_pdbx_mmcif,
+        "pymatgen": get_pymatgen_structure,
+    }
+
+    def __init__(self, structure: dict):
+        self._structure = None
+        self._converted_structures = dict.fromkeys(self._structure_type_converters)
+
+        self.structure = structure
+
+    @property
+    def structure(self) -> StructureResource:
+        """Get OPTIMADE structure"""
+        return self._structure
+
+    @structure.setter
+    def structure(self, value: dict):
+        """Set OPTIMADE structure
+        If already set, print that this can _only_ be set once.
+        """
+        if self._structure is None:
+            self._structure = StructureResource(**value)
+        else:
+            print("structure can only be set once and is already set.")
+
+    def __getattr__(self, name: str):
+        """Get converted structure or attribute from OPTIMADE structure
+        Order:
+        - Try to return converted structure if using `get_<valid_structures value>`.
+        - Try to return OPTIMADE StructureResource attribute.
```

Comment:
You are mixing namespaces here... is this really necessary?
What about defining a `def convert(self, format)` method that does the job?

Reply:
Nice idea with `convert()`.
Concerning the latter part - I'm not sure exactly what you mean? Would you have me sub-class `optimade.models.StructureResource`? I wanted explicitly _not_ to do this, in order to avoid getting the `pydantic` things and validators that followed along with it. But maybe it's fine, I'm not sure?

Reply:
Sorry for not being clearer. I simply meant to keep the second part of this function, which returns attributes of `self.structure`.